### PR TITLE
Forbid merge with null properties on a graph entity

### PR DESCRIPTION
**Pull request by Josipmrden**
Doing `MERGE ({id:null})` creates endless amounts of nodes. Same is done with edges if null property in the EdgeAtom. This is due to filter checking equality between 2 null values -> we don't have anything written in the PropertyStore since we treat nulls the same way as if the value is not present. Filter is checking then the property value of the filter null to be equal to the null of the storage, which is not because null == null does not hold true.
This is the reason why some of the nodes and edges get duplicated sometimes.
Solution to anyone using nulls in their system is to rewrite the query, i.e.

`MERGE (n) WITH n WHERE n.id is null SET ...`
[master < Task] PR
- [X] Check, and update documentation if necessary
- [X] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [X] Write a release note here, including added/changed clauses
- [X] Tag someone from docs team in the comments
@kgolubic 
Release note: "Merging node or edge with null property now raises an error."

DOCS PR: https://github.com/memgraph/documentation/pull/650

### CODE REVIEW

**Review by antepusic — APPROVED** [review]

**Comment by gitbuda** [comment]
@Josipmrden, does this work if `null` is injected as a parameter? It seems to be that also during the execution under `MERGE` the null value has to be checked? NOTE: This is not wrong. The check should be done in both places, because why even start the execution if you know the query is wrong (it's 100% possible to know) 🤔

File: tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature [hunk 1]
```diff
@@ -417,3 +417,93 @@ Feature: Merge feature
...
+    Scenario: Merge edge with unwind with null property error
+        Given an empty graph
+        When executing query:
+            """
+            UNWIND [1, 2, null, 3] as x
+            MERGE ()-[:TYPE {id:x}]->()
+            """
+        Then an error should be raised
+
+    Scenario: Merge node with null in on match passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature:463]:
```suggestion
    Scenario: Merge node with null in properties on match passes
```

File: tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature [hunk 2]
```diff
@@ -417,3 +417,93 @@ Feature: Merge feature
...
+            """
+        When executing query:
+            """
+            MERGE (n {a: 1})
+            ON MATCH SET n.prop = null
+            ON CREATE SET n.prop = null;
+            """
+        Then the result should be empty
+
+    Scenario: Merge edge with null in on match passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature:477]:
```suggestion
    Scenario: Merge edge with null in properties on match passes
```

File: tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature [hunk 3]
```diff
@@ -417,3 +417,93 @@ Feature: Merge feature
...
+            """
+        When executing query:
+            """
+            MERGE ()-[r:TYPE {a: 1}]->()
+            ON MATCH SET r.prop = null
+            ON CREATE SET r.prop = null;
+            """
+        Then the result should be empty
+
+    Scenario: Merge node with null in on create passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature:491]:
```suggestion
    Scenario: Merge node with null in properties on create passes
```

File: tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature [hunk 4]
```diff
@@ -417,3 +417,93 @@ Feature: Merge feature
...
+        Given an empty graph
+        When executing query:
+            """
+            MERGE (n {a: 1})
+            ON MATCH SET n.prop = null
+            ON CREATE SET n.prop = null;
+            """
+        Then the result should be empty
+
+    Scenario: Merge edge with null in on create passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1_on_disk/features/merge.feature:501]:
```suggestion
    Scenario: Merge edge with null in properties on create passes
```

File: tests/gql_behave/tests/memgraph_V1/features/merge.feature [hunk 5]
```diff
@@ -417,4 +417,93 @@ Feature: Merge feature
...
+        Given an empty graph
+        When executing query:
+            """
+            MERGE (n {a: 1})
+            ON MATCH SET n.prop = null
+            ON CREATE SET n.prop = null;
+            """
+        Then the result should be empty
+
+    Scenario: Merge edge with null in on create passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1/features/merge.feature:501]:
```suggestion
    Scenario: Merge edge with null in properties on create passes
```

File: tests/gql_behave/tests/memgraph_V1/features/merge.feature [hunk 6]
```diff
@@ -417,4 +417,93 @@ Feature: Merge feature
...
+            """
+        When executing query:
+            """
+            MERGE ()-[r:TYPE {a: 1}]->()
+            ON MATCH SET r.prop = null
+            ON CREATE SET r.prop = null;
+            """
+        Then the result should be empty
+
+    Scenario: Merge node with null in on create passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1/features/merge.feature:491]:
```suggestion
    Scenario: Merge node with null in properties on create passes
```

File: tests/gql_behave/tests/memgraph_V1/features/merge.feature [hunk 7]
```diff
@@ -417,4 +417,93 @@ Feature: Merge feature
...
+    Scenario: Merge edge with unwind with null property error
+        Given an empty graph
+        When executing query:
+            """
+            UNWIND [1, 2, null, 3] as x
+            MERGE ()-[:TYPE {id:x}]->()
+            """
+        Then an error should be raised
+
+    Scenario: Merge node with null in on match passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1/features/merge.feature:463]:
```suggestion
    Scenario: Merge node with null in properties on match passes
```

File: tests/gql_behave/tests/memgraph_V1/features/merge.feature [hunk 8]
```diff
@@ -417,4 +417,93 @@ Feature: Merge feature
...
+            """
+        When executing query:
+            """
+            MERGE (n {a: 1})
+            ON MATCH SET n.prop = null
+            ON CREATE SET n.prop = null;
+            """
+        Then the result should be empty
+
+    Scenario: Merge edge with null in on match passes
```
Comment (antepusic) [inline tests/gql_behave/tests/memgraph_V1/features/merge.feature:477]:
```suggestion
    Scenario: Merge edge with null in properties on match passes
```

**Review by antepusic — APPROVED** [review]
Looks good to me, only left small corrections to text naming

**Review by antepusic — APPROVED** [review]
Thanks for addressing the review comments, LGTM

**Pull request merged**